teps/0060-remote-resource-resolution.md

```diff
@@ -242,207 +251,533 @@ users can leverage those shared pipelines.
+kind: ResourceRequest
+metadata:
+  name: get-task-from-bundle-12345
+  namespace: bar
+  labels:
+    resolution.tekton.dev/resolver: bundle
+  ownerReferences:
+  - apiVersion: tekton.dev/v1beta1
+    controller: true
+    kind: TaskRun
+    name: my-tr
+    uid: 6aa5857a-3d67-4a09-94a1-8e9cc136dcf8
+spec:
+  params:
+    image_url: gcr.io/tekton-releases/catalog/upstream/golang-fuzz:0.1
+    name: golang-fuzz
+```
+
+A newly-created `ResourceRequest` for a task from the catalog git repo:
+
+```yaml
+apiVersion: resolution.tekton.dev/v1alpha1
+kind: ResourceRequest
+metadata:
+  name: get-task-from-git-12345
+  namespace: quux
+  labels:
+    resolution.tekton.dev/resolver: git
+  ownerReferences:
+  - apiVersion: tekton.dev/v1beta1
+    controller: true
+    kind: TaskRun
+    name: my-tr
+    uid: 6aa5857a-3d67-4a09-94a1-8e9cc136dcf8
+spec:
+  params:
+    repository_url: https://github.com/tektoncd/catalog.git
+    branch: main
+    path: /task/golang-fuzz/0.1/golang-fuzz.yaml
+```
+
+A successfully completed `ResourceRequest`:
+
+```yaml
+apiVersion: resolution.tekton.dev/v1alpha1
+kind: ResourceRequest
+metadata:
+  name: get-pipeline-from-git-12345
+  namespace: quux
+  labels:
+    resolution.tekton.dev/resolver: git
+  ownerReferences:
+  - apiVersion: tekton.dev/v1beta1
+    controller: true
+    kind: PipelineRun
+    name: my-pr
+    uid: 6aa5857a-3d67-4a09-94a1-8e9cc136dcf8
+spec:
+  params:
+    repository_url: https://github.com/sbwsg/experimental.git
+    commit: 2c9b093e94f30f588dc798cc56a2559d4da9d573
+    path: /remote-resolution/pipeline.yaml
+status:
+  annotations:
+    commit: 2c9b093e94f30f588dc798cc56a2559d4da9d573
+    content-type: application/x-yaml
+  data: a2luZDogUGlwZWxpbmUKYXBpVmVyc2lvbjogdGVrdG9uLmRldi92MWJldGExCm1ldGFkYXRhOgogIG5hbWU6IHAKc3BlYzoKICB0YXNrczoKICAtIG5hbWU6IGZldGNoLWZyb20tZ2l0CiAgICB0YXNrU3BlYzoKICAgICAgc3RlcHM6CiAgICAgIC0gaW1hZ2U6IGFscGluZS9naXQKICAgICAgICBzY3JpcHQ6IHwKICAgICAgICAgIGdpdCBjbG9uZSBodHRwczovL2dpdGh1Yi5jb20vdGVrdG9uY2QvcmVzdWx0cwo=
+  conditions:
+  - lastTransitionTime: "2021-10-14T15:38:14Z"
+    status: "True"
+    type: Succeeded
+```
+
+A failed `ResourceRequest`:
+
+```yaml
+apiVersion: resolution.tekton.dev/v1alpha1
+kind: ResourceRequest
+metadata:
+  name: get-pipeline-from-git-12345
+  namespace: quux
+  labels:
+    resolution.tekton.dev/resolver: git
+  ownerReferences:
+  - apiVersion: tekton.dev/v1beta1
+    controller: true
+    kind: PipelineRun
+    name: my-pr
+    uid: 6aa5857a-3d67-4a09-94a1-8e9cc136dcf8
+spec:
+  params:
+    repository_url: https://github.com/sbwsg/experimental.git
+    branch: remote-resolution
+    path: /remote-resolution/invalid-pipeline.yaml
+status:
+  conditions:
+  - lastTransitionTime: "2021-10-19T11:06:00Z"
+    message: 'error opening file "/remote-resolution/invalid-pipeline.yaml": file does not exist'
+    reason: ResolutionFailed
+    status: "False"
+    type: Succeeded
+```
+
+### Resolver specifics
```

Comment:
This will depend a bit on the requirements of each resolver. Taking `Git` as an example, during initial development my plan is to support configuration via a config file mounted e.g. from a `ConfigMap`.

The config for git resolution will allow an operator to specify global and per-repository settings that can limit (a) which repos are accessible _at all_ [^1] and (b) which repos are allowed from specific namespaces. Here's a bit of hand-waved config to center the discussion:

```yaml
allowed-repositories:

- name: tekton-catalog
  repository_url: https://github.com/tektoncd/catalog.git

- name: app-team-1-private-repo
  repository_url: git@code.enterprise-foo.net:app-team-1
  allowed-namespaces:
  - app-team-1
```

In this example config the operator has cut off access to all repos except for the public tekton catalog and an app team's private repo. Access to the app team's private repo is limited to `ResourceRequest` objects created in the `app-team-1` namespace. A git resolver can enforce this by virtue of the fact that `ResourceRequests` are namespace-scoped.

Looking a bit further forward, as part of the Workflows WG we've been discussing a `Repository CRD` object that could expose some or all of this configuration for use across multiple applications, not just by a git resolver but also, for example, by [a controller polling a repo for changes](https://github.com/tektoncd/community/pull/517#discussion_r727460016)).

[^1]: While the scope of this discussion is around repo access ideally the available config for git resolution is going to be more granular than that. E.g. let operators disallow resources pulled from any but a limited set of subdirectories in a repo, exact file paths within a repo, pulled from any but specific commit digests, or only from signed commits (with a specific set of authors' public keys).

Reply:
SGTM - I'd like to dig into the config syntax a bit more (I think we would benefit by leveraging a similar syntax as https://git-scm.com/docs/gitcredentials#_credential_contexts), but happy to punt this to another discussion to keep the TEP moving.